### DISCUSSION TITLE
Harpy for Twitter like site

### DISCUSSION THREAD

**Discussion by ghost** [robertodoering/harpy#536] [General]
Hello,
I want to learn android programing, I want to try to use Harpy as a front end of a difference social network, maybe for a mastodon instance, etc. Just to practice.

Do you have any recommendation where to start from?
Is the app decoupled from Twitter API?

A thing or two about the architecture of the projet would help a lot.

Kind regards,
Elijah

**Comment by robertodoering** [comment]
Hey,
first thing you should know that harpy is built using [Flutter](https://flutter.dev/) and is not a native Android app.

If you want to decouple the app from Twitter, you would have to go through each service and talk to whatever host you are interested in, instead of the Twitter API. Afterwards you can either map your data to the existing models that are used in harpy ([TweetData](https://github.com/robertodoering/harpy/blob/master/lib/api/twitter/data/tweet_data.dart), [UserData](https://github.com/robertodoering/harpy/blob/master/lib/api/twitter/data/user_data.dart), etc.) or replace them with your new models.

[flutter_bloc](https://pub.dev/packages/flutter_bloc) is used to manage state,
[get_it](https://pub.dev/packages/get_it) is used as a service locator and
[dart_twitter_api](https://pub.dev/packages/dart_twitter_api) is a package that I wrote to wrap the Twitter API and map its responses into dart models.

If you want to understand the architecture in harpy better it's best to make yourself familiar with flutter_bloc and state management in Flutter beforehand.
If you find anything you don't understand or have any question related to harpy and its architecture feel free to ask, I'm always happy to help.

**Reply by ghost** [reply]
Thanks a lot
